lib/SILOptimizer/Utils/OwnershipOptUtils.cpp

```diff
@@ -1051,3 +1121,247 @@ OwnershipRAUWHelper::perform(SingleValueInstruction *maybeTransformedNewValue) {
   OwnershipRAUWUtility utility{oldValue, actualNewValue, *ctx};
   return utility.perform();
 }
+
+//===----------------------------------------------------------------------===//
+//                           Single Use Replacement
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Given a use and a new value, lifetime extend new value as appropriate so we
+/// can replace use->get() with newValue and preserve ownership invariants. We
+/// assume that old value will be left alone and not deleted so we insert
+/// compensating cleanups.
+struct SingleUseReplacementUtility {
```

Comment:
I'll need to review `SingleUseReplacementUtility` later

Reply:
Its nothing too crazy.